##### **1. Базовые операции с векторами**

---
##### Задача 1.1 Нормы векторов (L1, L2)
**Вводные данные**: 

Вектор признаков объекта 

x = [120, 80, 200, 50]

Задача:
- [x] Реализовать функции вычисления L1 и L2 норм
- [x] Нормализовать вектор по L1 и по L2
- [ ] Сравнить, как меняется вклад каждого признака после нормализации

Что найти:
- L1(x), L2(x)
- нормализованные векторы
- интерпретацию разницы (устойчивость к выбросам)


---
##### Реализация функции вычисления L1 и L2 норм

**L1 (Манхэттенская норма)** - способ измерения длины вектора с помощью суммирования всех его абсолютных(число по модулю) значений.
Расстояние по сетке(вертикально или горизонатльно)

Формула: 
$$
\|x\|_1 = \sum_{i=1}^{n} |x_i|
$$

**L2 (Евклидова норма)** - 
Расстояние по сетке(по диагонали)

$$
\|x\|_2 = \sqrt{x_1^2 + x_2^2 + \dots + x_n^2} = \sqrt{\sum_{i=1}^{n} x_i^2}
$$

In [54]:
# input x
from math import sqrt


x = [120, 80, 200, 50]

#Функция вычисления Манхэттенской нормы
def L1(input_data):
    return sum(elem for elem in input_data)

#Функция вычисления Эвклидовой нормы
def L2(input_data):
    return sqrt(sum(elem**2 for elem in input_data))

print(f"Норма L1: {L1(x)}")
print(f"Норма L2: {L2(x)}")

Норма L1: 450
Норма L2: 251.5949125081825


---
##### Нормализация вектора по L1 и по 

**Нормализация** - позволяет убрать влияние маштаба и сделать признаки сопоставимыми.
Ищется путем деления каждого элемента на норму

**L1 нормализация** - используются, чтобы оценить вклад каждого элемента

**L2 нормализация** - оставляет только направление, показывая вклад каждого признака в длину вектора

In [55]:
input_data = x

# Норма L1
l1_norm = [elem/L1(x) for elem in input_data]

# Норма L2
l2_norm = [elem/L2(x) for elem in input_data]

print(f"Нормализация L1: {l1_norm}")
print(f"Нормализация L2: {l2_norm}")

Нормализация L1: [0.26666666666666666, 0.17777777777777778, 0.4444444444444444, 0.1111111111111111]
Нормализация L2: [0.47695718011029853, 0.3179714534068657, 0.7949286335171643, 0.19873215837929106]


---
##### Сравнение влияния признаков после нормализации

**После L1 нормализации**: 

Было **[120, 80, 200, 50]**, стало **[0.26666666666666666, 0.17777777777777778, 0.4444444444444444, 0.1111111111111111]**


**После L2 нормализации**: 

Было **[120, 80, 200, 50]**, стало **[0.47695718011029853, 0.3179714534068657, 0.7949286335171643, 0.19873215837929106]**

Сильнее величивает величину признаков(из-за квадрата в формуле), чем при L1.

##### **Вывод**: оба способа нормализации делают признаки устойчивым к масштабу, также увеличивают величину признака пропорционально его вклада. Но каждый способ имеет свои особенности.